text/0000-proc-macros.md

```diff
@@ -0,0 +1,417 @@
+to summarise: effectively, naming takes place in the context of the macro
+definition, not the expanded macro.
+
+Procedural macros often want to bend the rules around macro hygiene, for example
+to make items or variables more widely nameable than they would be by default.
+Procedural macros will be able to take part in the application of the hygiene
+algorithm via libmacro. Again, full details must wait for the libmacro RFC and a
+sketch is available in this [blog post](http://ncameron.org/blog/libmacro/).
+
+
+## Tokens
+
+Procedural macros will primarily operate on tokens. There are two main benefits
+to this principal: flexibility and future proofing. By operating on tokens, code
+passed to procedural macros does not need to satisfy the Rust parser, only the
+lexer. Stabilising an interface based on tokens means we need only commit to
+not changing the rules around those tokens, not the whole grammar. I.e., it
+allows us to change the Rust grammar without breaking procedural macros.
+
+In order to make the token-based interface even more flexible and future-proof,
+I propose a simpler token abstraction than is currently used in the compiler.
+The proposed system may be used directly in the compiler or may be an interface
+wrapper over a more efficient representation.
+
+Since macro expansion will not operate purely on tokens, we must keep hygiene
+information on tokens, rather than on `Ident` AST nodes (we might be able to
+optimise by not keeping such info for all tokens, but that is an implementation
+detail). We will also keep span information for each token, since that is where
+a record of macro expansion is maintained (and it will make life easier for
+tools. Again, we might optimise internally).
+
+A token is a single lexical element, for example, a numeric literal, a word
+(which could be an identifier or keyword), a string literal, or a comment.
+
+A token stream is a sequence of tokens, e.g., `a b c;` is a stream of four
+tokens - `['a', 'b', 'c', ';'']`.
+
+A token tree is a tree structure where each leaf node is a token and each
+interior node is a token stream. I.e., a token stream which can contain nested
+token streams. A token tree can be delimited, e.g., `a (b c);` will give
+`TT(None, ['a', TT(Some('()'), ['b', 'c'], ';'']))`. An undelimited token tree
+is useful for grouping tokens due to expansion, without representation in the
+source code. That could be used for unsafety hygiene, or to affect precedence
+and parsing without affecting scoping. They also replace the interpolated AST
+tokens currently in the compiler.
+
+In code:
+
+```
+// We might optimise this representation
+pub struct TokenStream(Vec<TokenTree>);
+
+// A borrowed TokenStream
+pub struct TokenSlice<'a>(&'a [TokenTree]);
+
+// A token or token tree.
+pub struct TokenTree {
+    pub kind: TokenKind,
+    pub span: Span,
+    pub hygiene: HygieneObject,
+}
+
+pub enum TokenKind {
+    Sequence(Delimiter, Vec<TokenTree>),
```

Comment:
We have to tokenise into sequences to parse patterns for macros, so I think the compiler must do it. It is then convenient for macro authors to have that info. Since we can't be more flexible by not doing it, I don't think there is an advantage of keeping that info private. Furthermore, it seems like a win for the macro author to see what the compiler sees in terms of why it is being passed the data it is.

Finally, using un-delimited sequences is useful for macros to affect precedence without introducing scopes.